Azure OpenAI: support audio transcription model

SandraAhlgrimm opened the pull request:
Hi! I have started with the audio transcription and would like to get feedback to ensure I'm heading in the right direction, as this is my first PR for this project.

## Issue
Closes #255

## Change
Add Azure OpenAI

## General checklist
<!-- Please double-check the following points and mark them like this: [X] -->
- [X] There are no breaking changes
- [X] I have added unit and integration tests for my change
- [X] I have manually run all the unit and integration tests in the module I have added/changed, and they are all green
- [X] I have manually run all the unit and integration tests in the [core](https://github.com/langchain4j/langchain4j/tree/main/langchain4j-core) and [main](https://github.com/langchain4j/langchain4j/tree/main/langchain4j) modules, and they are all green
<!-- Before adding documentation and example(s) (below), please wait until the PR is reviewed and approved. -->
- [x] I have added/updated the [documentation](https://github.com/langchain4j/langchain4j/tree/main/docs/docs)
- [x] I have added an example in the [examples repo](https://github.com/langchain4j/langchain4j-examples) (only for "big" features)
- [ ] I have added/updated [Spring Boot starter(s)](https://github.com/langchain4j/langchain4j-spring) (if applicable)

SandraAhlgrimm committed:
wip

SandraAhlgrimm committed:
wip

SandraAhlgrimm committed:
merge

SandraAhlgrimm committed:
Merge branch 'langchain4j:main' into add-azure-openai-audio

SandraAhlgrimm committed:
Merge branch 'main' of https://github.com/SandraAhlgrimm/langchain4j into add-azure-openai-audio

SandraAhlgrimm committed:
Merge branch 'langchain4j:main' into add-azure-openai-audio

SandraAhlgrimm committed:
cleanup

SandraAhlgrimm committed:
cleanup

SandraAhlgrimm committed:
Merge branch 'add-azure-openai-audio' of https://github.com/SandraAhlgrimm/langchain4j into add-azure-openai-audio

SandraAhlgrimm committed:
add custom handlers

SandraAhlgrimm committed:
add custom handlers

SandraAhlgrimm committed:
add audio file

SandraAhlgrimm committed:
add audio file

SandraAhlgrimm committed:
add audio

SandraAhlgrimm committed:
add audio

SandraAhlgrimm committed:
no base64 needed for Azure Open AI

SandraAhlgrimm committed:
no base64 needed for Azure Open AI

SandraAhlgrimm committed:
cleanup

SandraAhlgrimm committed:
cleanup

SandraAhlgrimm committed:
cleanup

omarmahamid commented on langchain4j-azure-open-ai/pom.xml:
```diff
@@ -147,6 +147,12 @@
+        
+        <dependency>
+            <groupId>commons-io</groupId>
+            <artifactId>commons-io</artifactId>
```
this used only in tests

make the scope for testing

omarmahamid commented on langchain4j-azure-open-ai/src/main/java/dev/langchain4j/model/azure/AzureOpenAiAudioModel.java:
```diff
@@ -0,0 +1,411 @@
+    private static final Logger logger = LoggerFactory.getLogger(AzureOpenAiAudioModel.class);
+    private OpenAIClient client;
+    private final String deploymentName;
+    private Audio audio = null;
```
isn't the interface got Audio model ?

Response<String> transcribe(Audio audioFile);

what is the need of the model to be stateful for audio ?
or maybe I'm missing something

omarmahamid commented on langchain4j-azure-open-ai/src/main/java/dev/langchain4j/model/azure/AzureOpenAiAudioModel.java:
```diff
@@ -0,0 +1,411 @@
+    private final String deploymentName;
+    private Audio audio = null;
+    private String prompt = null;
+    private String user = null;
```
for which purpose the user used ?

SandraAhlgrimm replied:
for managed identities

omarmahamid commented on langchain4j-azure-open-ai/src/main/java/dev/langchain4j/model/azure/AzureOpenAiAudioModel.java:
```diff
@@ -0,0 +1,411 @@
+    private String language;
+
+
+    //By default, the response type will be json with the raw text included.
```
I think this comment should not be in the ctor

omarmahamid commented on langchain4j-core/src/main/java/dev/langchain4j/data/audio/Audio.java:
```diff
@@ -140,4 +163,19 @@ public Audio build() {
     }
+
+    /**
+     * Get the filename of the audio using a regex to retrieve the last segment after the last slash.
```
I think the word regex isn't really used

SandraAhlgrimm replied:
You want me to call it Regular Expression instead?

omarmahamid replied:
No

1. You dont use regex to retrieve the last segment
2. You hardcoded the "/" and not use File.Separator (different os)

omarmahamid commented on langchain4j-core/src/main/java/dev/langchain4j/data/audio/Audio.java:
```diff
@@ -140,4 +163,19 @@ public Audio build() {
+            String path = url.getPath();
+            if (path != null) {
+                String[] segments = path.split("/");
+                return segments[segments.length - 1];
```
segments can't be zero elements ?

SandraAhlgrimm replied:
good point, thank you

SandraAhlgrimm committed:
enable filename retrival on root level, clean dependencies as requested by review

SandraAhlgrimm committed:
enable filename retrival on root level, clean dependencies as requested by review

SandraAhlgrimm committed:
move comment to constructor as requested by review

SandraAhlgrimm committed:
move comment to constructor as requested by review

agoncal commented:
@SandraAhlgrimm  any help on this one ?

SandraAhlgrimm commented:
> @SandraAhlgrimm any help on this one ?

thanks, I put a meeting in your calendar

243006306 commented:
When will this be completed?

243006306 commented:
Is this feature not included in subsequent releases? I think this feature is very useful with ASR

anunnakian commented on langchain4j-azure-open-ai/src/main/java/dev/langchain4j/model/azure/AzureOpenAiAudioModel.java:
```diff
@@ -0,0 +1,411 @@
+ * client secret of the AAD application as environment variables: AZURE_CLIENT_ID, AZURE_TENANT_ID, AZURE_CLIENT_SECRET.
+ * Then, provide the DefaultAzureCredential instance to the builder: `builder.tokenCredential(new DefaultAzureCredentialBuilder().build())`.
+ */
+public class AzureOpenAiAudioModel implements AudioModel {
```
Yes, we should differentiate Transcription API between others Open AI APIs (Speech API and Translation API) @SandraAhlgrimm I can help on that if you want :)

anunnakian commented on langchain4j-azure-open-ai/src/test/java/dev/langchain4j/model/azure/AzureOpenAiAudioModelIT.java:
```diff
@@ -0,0 +1,43 @@
+        Audio audio = Audio.builder()
+                 .audioData(audioData)
+                 .mimeType("audio/mpeg")
+                 .url(audioFile.toURI())
```
The Vanilla OpenAI Transcription API expected a file as a multipart parameter in the request (not a json). If you check the documentation it says that "The audio file object (not file name) to transcribe"

If I understand the last part of your comment, you are proposing to set a file as a default, and each time we want to transcribe an audio, we replace this file and then call the transcription API ?

anunnakian commented on langchain4j-azure-open-ai/src/main/java/dev/langchain4j/model/azure/AzureOpenAiAudioModelName.java:
```diff
@@ -0,0 +1,40 @@
+package dev.langchain4j.model.azure;
+
+public enum AzureOpenAiAudioModelName {
```
I think we should remove this class an specify the model name as a string

anunnakian commented on langchain4j-core/src/main/java/dev/langchain4j/data/audio/Audio.java:
```diff
@@ -10,6 +10,7 @@
 @Experimental
 public class Audio {
     private final URI url;
+    private final byte[] audioData;
```
What do you think @langchain4j if we move this class into the azure openai lc4j module ? this object corresponds to Azure Open AI case

SandraAhlgrimm replied:
@anunnakian The Audio class is generic, not Azure-specific:

It's in the [langchain4j-core](https://humble-dollop-vwr6q75gp72rv6.github.dev/) module under dev.langchain4j.data.audio
It's part of the core audio abstraction alongside AudioTranscriptionRequest and AudioTranscriptionResponse
The AudioModel interface in core depends on it
It's designed for reusability:

Any future audio model implementation (OpenAI, Google, AWS, etc.) would need this class
Moving it to Azure module would create circular dependencies or force duplication
Current architecture is correct:

Core abstractions (Audio, AudioModel, Request/Response) belong in [langchain4j-core]
Implementation-specific code belongs in provider modules (azure-open-ai, openai, etc.)

anunnakian commented on langchain4j-core/src/main/java/dev/langchain4j/model/audio/AudioModel.java:
```diff
@@ -0,0 +1,14 @@
+import dev.langchain4j.data.audio.Audio;
+import dev.langchain4j.model.output.Response;
+
+public interface AudioModel {
```
This class as well, it should be located in the Azure OpenAi lc4j module

anunnakian committed:
Review comments

anunnakian committed:
Revert Audio class

anunnakian committed:
Use the Audio class in the core

anunnakian commented:
@LizeRaes @langchain4j @dliubarskyi any thoughts :) ?

anunnakian committed:
Merge branch 'main' into add-azure-openai-audio

SandraAhlgrimm committed:
add documentation for audio transcription

dliubarskyi commented:
@anunnakian sorry, thoughts on what exactly? Could you please elaborate?

anunnakian commented:
> @anunnakian sorry, thoughts on what exactly? Could you please elaborate?

Hi @dliubarskyi, I took care of all comments and change requests about this PR (@SandraAhlgrimm gave me access to her branch). Could you please review it 🙏 
I'm ready to handle all your feedbacks quickly

SandraAhlgrimm committed:
Resolve merge conflicts by preferring upstream/main changes in Audio.java and pom.xml

SandraAhlgrimm committed:
Update Azure OpenAI model dependencies and improve audio transcription test

SandraAhlgrimm committed:
Merge branch 'add-azure-openai-audio' of https://github.com/SandraAhlgrimm/langchain4j into add-azure-openai-audio

SandraAhlgrimm committed:
Fix duplicate dependencies in Azure OpenAI module

SandraAhlgrimm committed:
Address PR feedback: Refactor Azure OpenAI Audio API to use request/response pattern, fix path handling, and clean up implementation

SandraAhlgrimm committed:
Enhance Audio class documentation and fix equals() method to include audioData field

SandraAhlgrimm committed:
Fix deprecation warnings in AzureOpenAiAudioTranscriptionModelIT by adding @SuppressWarnings and documentation

SandraAhlgrimm committed:
Fix SLF4J multiple bindings issue by explicitly configuring Log4j2 and excluding TinyLog

SandraAhlgrimm committed:
Refactor audio processing: remove deprecated methods, enhance AudioModel interface, and add comprehensive tests for Audio.getFilename() functionality

SandraAhlgrimm commented on langchain4j-azure-open-ai/src/main/java/dev/langchain4j/model/azure/AzureOpenAiAudioModelName.java:
```diff
@@ -0,0 +1,40 @@
+    WHISPER("whisper-1", "whisper-1"); // alias for the latest whisper-2large model
+
+    private final String modelName;
+    // Model type follows the com.knuddels.jtokkit.api.ModelType naming convention
```
jtokkit is designed for tokenizing text, not audio. Audio files are binary data that gets processed differently than text tokens. Thanks for the hint. I removed it/

SandraAhlgrimm commented:
Hi @jdubois and @agoncal, I am looking forward to your review

jdubois commented:
Spotless is modifying a lot of files that have nothing to do with this PR, I know it will make the CI/CD break but I wouldn't commit those here.

SandraAhlgrimm committed:
Add Azure OpenAI audio transcription functionality

- Add AzureOpenAiAudioTranscriptionModel with comprehensive builder pattern
- Support for Whisper model integration with Azure OpenAI
- Include comprehensive test coverage with AzureOpenAiAudioTranscriptionModelIT
- Add SPI factory for dependency injection support
- Include test audio resource (hello.mp3) for integration testing
- Clean implementation focused only on audio functionality without unrelated formatting changes

SandraAhlgrimm committed:
Merge branch 'main' into add-azure-openai-audio

SandraAhlgrimm commented:
@jdubois, I took spotless fixes out as discussed and will file this in another PR.

It's ready for your review

SandraAhlgrimm commented:
Issue #3576
Pull Request #3577 for formatting

dliubarskyi commented on docs/docs/integrations/language-models/azure-open-ai.md:
```diff
@@ -342,7 +342,68 @@ langchain4j.azure-open-ai.streaming-chat-model.user-agent-suffix=...
+<dependency>
+    <groupId>dev.langchain4j</groupId>
+    <artifactId>langchain4j-azure-open-ai</artifactId>
+    <version>1.0.1-beta6</version>
```
```suggestion
    <version>1.4.0</version>
```

dliubarskyi commented on langchain4j-azure-open-ai/pom.xml:
```diff
@@ -43,17 +43,58 @@
+
+        <dependency>
+            <groupId>com.azure</groupId>
+            <artifactId>azure-identity</artifactId>
```
This dependency was removed some time ago, I guess it is an artifact of a merge?

dliubarskyi commented on langchain4j-azure-open-ai/pom.xml:
```diff
@@ -64,6 +105,10 @@
 
+        <dependency>
+            <groupId>org.slf4j</groupId>
+            <artifactId>slf4j-api</artifactId>
```
this dependency should not be required

dliubarskyi commented on langchain4j-azure-open-ai/src/main/java/dev/langchain4j/model/azure/AzureOpenAiAudioTranscriptionModel.java:
```diff
@@ -0,0 +1,330 @@
+     * @param deploymentName The deployment name of the audio model
+     * @param responseFormat The response format (can be null for default JSON format)
+     */
+    public AzureOpenAiAudioTranscriptionModel(
```
It would be nice to have a constructor that accepts only `Builder`, similarly to `AzureOpenAiChatModel`. Otherwise, it will be harder to add new parameters without breaking change or introducing more and more ctors in the future

dliubarskyi commented on langchain4j-core/src/main/java/dev/langchain4j/model/audio/AudioModel.java:
```diff
@@ -0,0 +1,45 @@
+     * @deprecated This method will be removed in a future release. Use {@link #transcribe(AudioTranscriptionRequest)} 
+     * for full control or {@link #transcribeToText(Audio)} for a simpler API that returns just the text.
+     */
+    @Deprecated
```
This is a new class, deprecating does not make sense, it can be just removed

dliubarskyi commented on langchain4j-core/src/main/java/dev/langchain4j/data/audio/Audio.java:
```diff
@@ -5,9 +5,25 @@
 public class Audio {
 
     private final URI url;
+    private final byte[] audioData;
```
I would rename it into `binaryData` to be more specific and to be on the same page with `base64Data`. We are inside the `Audio` class, so "audio" prefix does not bring any new information, but the binary nature of this data is missing in the name.

dliubarskyi commented on langchain4j-azure-open-ai/src/main/java/dev/langchain4j/model/azure/AzureOpenAiAudioTranscriptionModel.java:
```diff
@@ -0,0 +1,330 @@
+                .setFilename(filename)
+                .setResponseFormat(responseFormat);
+
+        // Only set language if provided
```
Please remove all uninformative comments

dliubarskyi reviewed (commented):
@SandraAhlgrimm thanks a lot! 🙏

dliubarskyi commented:
@anunnakian so sorry about the delay with the review!

SandraAhlgrimm committed:
Address review comments for Azure OpenAI audio transcription

- Update documentation version from 1.0.1-beta6 to 1.4.0
- Clean up POM dependencies: remove jackson-annotations, move slf4j to test scope
- Add constructor accepting only Builder parameter
- Remove deprecated transcribe(Audio) method from AudioModel interface
- Rename audioData to binaryData in Audio class for consistency
- Remove uninformative getter/setter comments
- Update tests to use new binaryData() method

All tests passing with Azure OpenAI environment variables.

SandraAhlgrimm committed:
fix: Add both langchain4j regular and test-jar dependencies

The Azure OpenAI module needs:
- Regular langchain4j dependency for runtime code access
- Test-jar dependency for common test base classes (AbstractAiServiceIT, etc.)

This fixes the compilation error for the service integration tests.

SandraAhlgrimm commented:
@dliubarskyi thanks for your speedy review! I addressed your comments.

SandraAhlgrimm committed:
Merge branch 'main' into add-azure-openai-audio

SandraAhlgrimm committed:
complete merge

SandraAhlgrimm committed:
Merge remote-tracking branch 'upstream/main' into add-azure-openai-audio

SandraAhlgrimm committed:
Merge remote-tracking branch 'upstream/main' into add-azure-openai-audio

dliubarskyi commented on docs/docs/integrations/language-models/azure-open-ai.md:
```diff
@@ -342,7 +342,68 @@ langchain4j.azure-open-ai.streaming-chat-model.user-agent-suffix=...
+    .deploymentName("your-audio-model-deployment-name") // e.g., "whisper"
+    .build();
+
+String transcript = model.transcribe(new File("path/to/audio-file.wav"));
```
There is no `transcribe(File)` method

dliubarskyi commented on docs/docs/integrations/language-models/azure-open-ai.md:
````diff
@@ -342,7 +342,68 @@ langchain4j.azure-open-ai.streaming-chat-model.user-agent-suffix=...
+System.out.println(transcript);
+```
+
+### Spring Boot Usage
````
I would remove this section as there is no SB implementation yet

dliubarskyi commented on langchain4j-azure-open-ai/src/main/java/dev/langchain4j/model/azure/AzureOpenAiAudioTranscriptionModel.java:
```diff
@@ -0,0 +1,346 @@
+            filename = "audio.mp3";
+        }
+
+        AudioTranscriptionOptions options = new AudioTranscriptionOptions(audio.binaryData())
```
It is worth checking that `audio.binaryData()` is not `null`, as `Audio` can be created from other source (e.g., URL or base64)

dliubarskyi commented on langchain4j-azure-open-ai/src/test/java/dev/langchain4j/model/azure/AzureOpenAiAudioTranscriptionModelIT.java:
```diff
@@ -0,0 +1,108 @@
+        String deploymentName = System.getenv("AZURE_OPENAI_AUDIO_DEPLOYMENT_NAME");
+
+        // Skip test if environment variables aren't configured
+        Assumptions.assumeTrue(
```
nit: it is better to use `@EnabledIfEnvironmentVariable(named = "AZURE_OPENAI_ENDPOINT", matches = ".+")` on top of the class instead, you can see `AzureOpenAiChatModelIT` as an example

dliubarskyi commented on langchain4j-azure-open-ai/src/test/java/dev/langchain4j/model/azure/AzureOpenAiAudioTranscriptionModelIT.java:
```diff
@@ -0,0 +1,108 @@
+
+    @Test
+    void should_transcribe_audio_with_new_api() throws IOException {
+        String endpoint = System.getenv("AZURE_OPENAI_ENDPOINT");
```
```suggestion
        String endpoint = System.getenv("AZURE_OPENAI_AUDIO_ENDPOINT");
```
Do you mind having a separate variable? I was not able to deploy whisper to the same region as other Azure OpenAI models

dliubarskyi commented on langchain4j-azure-open-ai/src/test/java/dev/langchain4j/model/azure/AzureOpenAiAudioTranscriptionModelIT.java:
```diff
@@ -0,0 +1,108 @@
+    @Test
+    void should_transcribe_audio_with_new_api() throws IOException {
+        String endpoint = System.getenv("AZURE_OPENAI_ENDPOINT");
+        String apiKey = System.getenv("AZURE_OPENAI_KEY");
```
```suggestion
        String apiKey = System.getenv("AZURE_OPENAI_AUDIO_KEY");
```

dliubarskyi commented on langchain4j-azure-open-ai/src/test/java/dev/langchain4j/model/azure/AzureOpenAiAudioTranscriptionModelIT.java:
```diff
@@ -0,0 +1,108 @@
+                .build();
+
+        File audioFile = new File("src/test/resources/hello.mp3");
+        Assumptions.assumeTrue(audioFile.exists(), "Test audio file not found: src/test/resources/hello.mp3");
```
If there is no file, the test should fail instead of being ignored

dliubarskyi commented on langchain4j-azure-open-ai/src/test/java/dev/langchain4j/model/azure/AzureOpenAiAudioTranscriptionModelIT.java:
```diff
@@ -0,0 +1,108 @@
+        String deploymentName = System.getenv("AZURE_OPENAI_AUDIO_DEPLOYMENT_NAME");
+
+        // Skip test if environment variables aren't configured
+        Assumptions.assumeTrue(
```
same here

dliubarskyi commented on langchain4j-azure-open-ai/src/test/java/dev/langchain4j/model/azure/AzureOpenAiAudioTranscriptionModelIT.java:
```diff
@@ -0,0 +1,108 @@
+                .build();
+
+        File audioFile = new File("src/test/resources/hello.mp3");
+        Assumptions.assumeTrue(audioFile.exists(), "Test audio file not found: src/test/resources/hello.mp3");
```
same here

dliubarskyi commented on langchain4j-core/src/main/java/dev/langchain4j/data/audio/Audio.java:
```diff
@@ -64,38 +94,38 @@ public boolean equals(Object o) {
     public static class Builder {
 
-        private URI url;
+        URI url;
```
```suggestion
        private URI url;
```

dliubarskyi commented on langchain4j-core/src/main/java/dev/langchain4j/data/audio/Audio.java:
```diff
@@ -149,4 +194,43 @@ public Audio build() {
+     * If the audio data is a URL, return the name of the URL file.
+     * @return the file name.
+     */
+    public String getFilename() {
```
Why is this needed? As far as I understand, `Audio` should be created from the binary data, so filename will always be null and this method is redundant?

dliubarskyi replied:
Please address this comment, I guess this method should be removed

dliubarskyi commented on langchain4j-core/src/main/java/dev/langchain4j/model/audio/AudioModel.java:
```diff
@@ -0,0 +1,31 @@
+/**
+ * A model that can process audio inputs.
+ */
+public interface AudioModel {
```
```suggestion
public interface AudioTranscriptionModel {
```

dliubarskyi replied:
```suggestion
@Experimental
public interface AudioModel {
```

dliubarskyi commented on langchain4j-core/src/main/java/dev/langchain4j/model/audio/AudioModel.java:
```diff
@@ -0,0 +1,31 @@
+import dev.langchain4j.data.audio.Audio;
+
+/**
+ * A model that can process audio inputs.
```
```suggestion
 * A model that can transcribe audio into text.
```

dliubarskyi reviewed (commented):
@SandraAhlgrimm thanks a lot, there are a few minor comments left and we can merge!

dliubarskyi commented on langchain4j-azure-open-ai/src/main/java/dev/langchain4j/model/azure/AzureOpenAiAudioTranscriptionModel.java:
```diff
@@ -0,0 +1,346 @@
+ *     with an appropriate credential implementation like {@code DefaultAzureCredential}.</li>
+ * </ol>
+ */
+public class AzureOpenAiAudioTranscriptionModel implements AudioModel {
```
```suggestion
@Experimental
public class AzureOpenAiAudioTranscriptionModel implements AudioModel {
```

dliubarskyi commented on langchain4j-core/src/main/java/dev/langchain4j/model/audio/AudioTranscriptionRequest.java:
```diff
@@ -0,0 +1,115 @@
+/**
+ * Request to transcribe audio.
+ */
+public class AudioTranscriptionRequest {
```
```suggestion
@Experimental
public class AudioTranscriptionRequest {
```

dliubarskyi commented on langchain4j-core/src/main/java/dev/langchain4j/model/audio/AudioTranscriptionResponse.java:
```diff
@@ -0,0 +1,35 @@
+/**
+ * Response containing the transcription of an audio file.
+ */
+public class AudioTranscriptionResponse {
```
```suggestion
@Experimental
public class AudioTranscriptionResponse {
```

dliubarskyi commented:
@SandraAhlgrimm I plan to do a release tomorrow, and this PR is a good candidate to be included. Would you be able to address the comments till tomorrow noon? If not, no worries, I can do it myself, just allow me to push to your repo

SandraAhlgrimm committed:
Refactor audio transcription model and update documentation from review

- Introduced `AudioTranscriptionModel` interface for audio transcription.
- Updated `AzureOpenAiAudioTranscriptionModel` to implement the new interface.
- Enhanced audio data handling with a new helper method.
- Revised integration tests to reflect changes in environment variable usage.
- Updated documentation for audio transcription usage.

dliubarskyi commented on langchain4j-azure-open-ai/src/main/java/dev/langchain4j/model/azure/AzureOpenAiAudioTranscriptionModel.java:
```diff
@@ -124,6 +128,29 @@ public AudioTranscriptionResponse transcribe(AudioTranscriptionRequest request)
+            throw new IllegalArgumentException("URL-based audio is not supported by Azure OpenAI transcription. Please provide audio as binary data or base64 encoded data.");
+        }
+        
+        throw new IllegalArgumentException("No audio data found. Audio must contain either binary data, base64 data, or URL.");
```
```suggestion
        throw new IllegalArgumentException("No audio data found. Audio must contain either binary data, base64 data");
```

SandraAhlgrimm committed:
Update langchain4j-azure-open-ai/src/main/java/dev/langchain4j/model/azure/AzureOpenAiAudioTranscriptionModel.java

Co-authored-by: Dmytro Liubarskyi <ljubarskij@gmail.com>

dliubarskyi commented:
@SandraAhlgrimm please ping me once this will be ready for merge

SandraAhlgrimm committed:
add unit tests to pass jacoco check

SandraAhlgrimm committed:
Merge branch 'add-azure-openai-audio' of https://github.com/SandraAhlgrimm/langchain4j into add-azure-openai-audio

SandraAhlgrimm committed:
Merge branch 'main' into add-azure-openai-audio

SandraAhlgrimm committed:
refactor: clean up whitespace and improve code readability in AzureOpenAiAudioTranscriptionModel and its tests with spotless

SandraAhlgrimm committed:
refactor: sportless, streamline import statements and improve code formatting across multiple files

SandraAhlgrimm commented:
@dliubarskyi ready!

dliubarskyi commented on docs/docs/integrations/language-models/azure-open-ai.md:
```diff
@@ -342,7 +342,71 @@ langchain4j.azure-open-ai.streaming-chat-model.user-agent-suffix=...
+// Create Audio object with binary data
+Audio audio = Audio.builder()
+    .binaryData(audioData)
+    .mimeType("audio/wav")
```
```suggestion
```
mime type is not used anywhere

dliubarskyi commented on docs/docs/integrations/language-models/azure-open-ai.md:
```diff
@@ -342,7 +342,71 @@ langchain4j.azure-open-ai.streaming-chat-model.user-agent-suffix=...
+// Create transcription request
+AudioTranscriptionRequest request = AudioTranscriptionRequest.builder()
+    .audio(audio)
+    .language("en") // optional
```
```suggestion
    .prompt("This is an audio file containing ...") // optional
    .language("en") // optional
```

SandraAhlgrimm committed:
Update docs/docs/integrations/language-models/azure-open-ai.md

Co-authored-by: Dmytro Liubarskyi <ljubarskij@gmail.com>

SandraAhlgrimm committed:
Update docs/docs/integrations/language-models/azure-open-ai.md

Co-authored-by: Dmytro Liubarskyi <ljubarskij@gmail.com>

dliubarskyi committed:
Update langchain4j-core/src/main/java/dev/langchain4j/data/audio/Audio.java

dliubarskyi commented:
@SandraAhlgrimm only a single concern left: https://github.com/langchain4j/langchain4j/pull/2110#discussion_r2300613557

SandraAhlgrimm commented:
> @SandraAhlgrimm only a single concern left: [#2110 (comment)](https://github.com/langchain4j/langchain4j/pull/2110#discussion_r2300613557)

Well, we can remove the FileName, but it might cause issues with (Azure) OpenAI for some types and requests

dliubarskyi commented:
@SandraAhlgrimm it just does not make sense as far as I understand. `Audio` should be created from `binaryData` or `base64Data`, not from the `url`. But `Audio.getFilename()` expects a `url`.

Since there is no way to pass the file name at the moment, I would remove the `getFilename()` method and always use "audio.mp3" as the file name in the Azure OpenAI request. WDYT?

SandraAhlgrimm commented:
> @SandraAhlgrimm it just does not make sense as far as I understand. `Audio` should be created from `binaryData` or `base64Data`, not from the `url`. But `Audio.getFilename()` expects a `url`.
> 
> Since there is no way to pass the file name at the moment, I would remove the `getFilename()` method and always use "audio.mp3" as the file name in the Azure OpenAI request. WDYT?

ok, I am with you! Thanks for explaining extra clear!

SandraAhlgrimm committed:
refactor: simplify filename handling in AzureOpenAiAudioTranscriptionModel and related tests

SandraAhlgrimm committed:
Merge branch 'add-azure-openai-audio' of https://github.com/SandraAhlgrimm/langchain4j into add-azure-openai-audio

SandraAhlgrimm committed:
spotless fix

SandraAhlgrimm commented:
> Oh, and one more thing: do we plan to fully support all `response_format` and `timestamp_granularities`?

I put this here to finally get the current one merged first :D https://github.com/langchain4j/langchain4j/issues/3604

dliubarskyi reviewed (approved):
@SandraAhlgrimm thanks a lot, great job! 🙏
